JUPNP Message Parsing Failure due to DocumentBuilder Namespace Reset on Android

I'm in the process of migrating [Dsub2000](https://github.com/paroj/DSub2000), a fork of [Dsub](https://github.com/daneren2005/Subsonic), from cling to jupnp library. I've encountered a problem when the jupnp starts failing message parsing after 20 parsed messages. 

I have put up [a minimal reproducible example](https://github.com/norohind/DSub2000/blob/27aa40e964362a3caae7ae858444a36d05b86ec1/app/src/androidTest/java/github/daneren2005/dsub/domain/JUPNPTest.java) in form of an instrumented test, it makes jupnp parse 20 xml messages, prints `--- Crash after this line ---` and calls *readBody* once again to capture the failure. Here's the log output of running this example when the problem occurs:
```
04-21 13:35:24.856 16154 16177 I System.out: --- Crash after this line ---
04-21 13:35:24.856 16154 16177 V ojts.SOAPActionProcesso: Reading body of github.daneren2005.dsub.domain.JUPNPTest$1@6cf33de for: (ActionInvocation) (Action, Arguments: 0) GetPositionInfo
04-21 13:35:24.856 16154 16177 V ojts.SOAPActionProcesso: ===================================== SOAP BODY BEGIN ============================================
04-21 13:35:24.856 16154 16177 I System.out: getBodyString
04-21 13:35:24.856 16154 16177 V ojts.SOAPActionProcesso: <?xml version="1.0" encoding="utf-8"?><s:Envelope s:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/" xmlns:s="http://schemas.xmlsoap.org/soap/envelope/"><s:Body><u:GetPositionInfoResponse xmlns:u="urn:schemas-upnp-org:service:AVTransport:1"><Track>1</Track><TrackDuration>0</TrackDuration><TrackMetaData>doesn't really matter</TrackMetaData><TrackURI>doesn't really matter</TrackURI><RelTime>0:00:00</RelTime><AbsTime>0:00:00</AbsTime><RelCount>2147483647</RelCount><AbsCount>2147483647</AbsCount></u:GetPositionInfoResponse></s:Body></s:Envelope>
04-21 13:35:24.856 16154 16177 V ojts.SOAPActionProcesso: -===================================== SOAP BODY END ============================================
04-21 13:35:24.856 16154 16177 I System.out: IsBodyNonEmptyString
04-21 13:35:24.856 16154 16177 I System.out: getBodyString
04-21 13:35:24.858 16154 16177 E TestRunner: failed: testSetCreated(github.daneren2005.dsub.domain.JUPNPTest)
04-21 13:35:24.858 16154 16177 E TestRunner: ----- begin exception -----
04-21 13:35:24.860 16154 16177 E TestRunner: org.jupnp.model.UnsupportedDataException: Can't transform message payload
04-21 13:35:24.860 16154 16177 E TestRunner:    at org.jupnp.transport.impl.SOAPActionProcessorImpl.readBody(SOAPActionProcessorImpl.java:164)
04-21 13:35:24.860 16154 16177 E TestRunner:    at github.daneren2005.dsub.domain.JUPNPTest.testSetCreated(JUPNPTest.java:46)
04-21 13:35:24.860 16154 16177 E TestRunner:    at java.lang.reflect.Method.invoke(Native Method)
04-21 13:35:24.860 16154 16177 E TestRunner:    at junit.framework.TestCase.runTest(TestCase.java:168)
04-21 13:35:24.860 16154 16177 E TestRunner:    at junit.framework.TestCase.runBare(TestCase.java:134)
04-21 13:35:24.860 16154 16177 E TestRunner:    at junit.framework.TestResult$1.protect(TestResult.java:115)
04-21 13:35:24.860 16154 16177 E TestRunner:    at androidx.test.internal.runner.junit3.AndroidTestResult.runProtected(AndroidTestResult.java:73)
04-21 13:35:24.860 16154 16177 E TestRunner:    at junit.framework.TestResult.run(TestResult.java:118)
04-21 13:35:24.860 16154 16177 E TestRunner:    at androidx.test.internal.runner.junit3.AndroidTestResult.run(AndroidTestResult.java:51)
04-21 13:35:24.860 16154 16177 E TestRunner:    at junit.framework.TestCase.run(TestCase.java:124)
04-21 13:35:24.860 16154 16177 E TestRunner:    at androidx.test.internal.runner.junit3.NonLeakyTestSuite$NonLeakyTest.run(NonLeakyTestSuite.java:62)
04-21 13:35:24.860 16154 16177 E TestRunner:    at junit.framework.TestSuite.runTest(TestSuite.java:243)
04-21 13:35:24.860 16154 16177 E TestRunner:    at junit.framework.TestSuite.run(TestSuite.java:238)
04-21 13:35:24.860 16154 16177 E TestRunner:    at androidx.test.internal.runner.junit3.DelegatingTestSuite.run(DelegatingTestSuite.java:97)
04-21 13:35:24.860 16154 16177 E TestRunner:    at androidx.test.internal.runner.junit3.AndroidTestSuite.run(AndroidTestSuite.java:65)
04-21 13:35:24.860 16154 16177 E TestRunner:    at androidx.test.internal.runner.junit3.JUnit38ClassRunner.run(JUnit38ClassRunner.java:115)
04-21 13:35:24.860 16154 16177 E TestRunner:    at org.junit.runners.Suite.runChild(Suite.java:128)
04-21 13:35:24.860 16154 16177 E TestRunner:    at org.junit.runners.Suite.runChild(Suite.java:27)
04-21 13:35:24.860 16154 16177 E TestRunner:    at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
04-21 13:35:24.860 16154 16177 E TestRunner:    at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
04-21 13:35:24.860 16154 16177 E TestRunner:    at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
04-21 13:35:24.860 16154 16177 E TestRunner:    at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
04-21 13:35:24.860 16154 16177 E TestRunner:    at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
04-21 13:35:24.860 16154 16177 E TestRunner:    at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
04-21 13:35:24.860 16154 16177 E TestRunner:    at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
04-21 13:35:24.860 16154 16177 E TestRunner:    at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
04-21 13:35:24.860 16154 16177 E TestRunner:    at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
04-21 13:35:24.860 16154 16177 E TestRunner:    at androidx.test.internal.runner.TestExecutor.execute(TestExecutor.java:67)
04-21 13:35:24.860 16154 16177 E TestRunner:    at androidx.test.internal.runner.TestExecutor.execute(TestExecutor.java:58)
04-21 13:35:24.860 16154 16177 E TestRunner:    at androidx.test.runner.AndroidJUnitRunner.onStart(AndroidJUnitRunner.java:446)
04-21 13:35:24.860 16154 16177 E TestRunner:    at android.app.Instrumentation$InstrumentationThread.run(Instrumentation.java:2402)
04-21 13:35:24.860 16154 16177 E TestRunner: Caused by: java.lang.RuntimeException: Response root element was not 'Envelope'
04-21 13:35:24.860 16154 16177 E TestRunner:    at org.jupnp.transport.impl.SOAPActionProcessorImpl.readBodyElement(SOAPActionProcessorImpl.java:232)
04-21 13:35:24.860 16154 16177 E TestRunner:    at org.jupnp.transport.impl.SOAPActionProcessorImpl.readBody(SOAPActionProcessorImpl.java:154)
04-21 13:35:24.860 16154 16177 E TestRunner:    ... 30 more
04-21 13:35:24.860 16154 16177 E TestRunner: ----- end exception -----
04-21 13:35:24.864 16154 16177 I TestRunner: finished: testSetCreated(github.daneren2005.dsub.domain.JUPNPTest)

org.jupnp.model.UnsupportedDataException: Can't transform message payload
at org.jupnp.transport.impl.SOAPActionProcessorImpl.readBody(SOAPActionProcessorImpl.java:164)
at github.daneren2005.dsub.domain.JUPNPTest.testSetCreated(JUPNPTest.java:46)
... 28 trimmed
Caused by: java.lang.RuntimeException: Response root element was not 'Envelope'
at org.jupnp.transport.impl.SOAPActionProcessorImpl.readBodyElement(SOAPActionProcessorImpl.java:232)
at org.jupnp.transport.impl.SOAPActionProcessorImpl.readBody(SOAPActionProcessorImpl.java:154)
... 30 more
```

## My investigation
Jupnp reuses a pre-created pool of platform specific implementation of `javax.xml.parsers.DocumentBuilder`, which are created with a platform specific implementation of `javax.xml.parsers.DocumentBuilderFactory`. An acquired *platform specific* instance of a factory method `.setNamespaceAware(true)` is being [called](https://github.com/jupnp/jupnp/blob/3.0.1/bundles/org.jupnp/src/main/java/org/jupnp/transport/impl/PooledXmlProcessor.java#L57). It tells the factory to create instances of `DocumentBuilder` which are *namespace aware*. After every use of an instance of `DocumentBuilder` and before returning it to the pool, the jupnp [calls](https://github.com/jupnp/jupnp/blob/3.0.1/bundles/org.jupnp/src/main/java/org/jupnp/transport/impl/PooledXmlProcessor.java#L136) a `.reset()` method on the instance. 

I have checked on my physical phone and an emulator, this *platform specific* implementation is `org.apache.harmony.xml.parsers`. In this particular implementation, on method `.reset()` it unconditionally sets internal flag `namespaceAware` of a specific instance to false.

Jupnp ends up putting this not anymore *namespace aware* `DocumentBuilder` back to pool. Eventually (after 20 message parsed messages, this is default pool size) not *namespace aware* builder parses a message and in [this condition](https://github.com/jupnp/jupnp/blob/main/bundles/org.jupnp/src/main/java/org/jupnp/transport/impl/SOAPActionProcessorImpl.java#L232) jupnp gets `s:Envelope` `getUnprefixedNodeName(envelopeElement)` as return value of the call and throws RuntimeException line below.

*Namespace awareness* is important, because it allows the parser to distinguish prefixes in xml tags, like prefix `s` and in tag `Envelope` in raw tag xml element `s:Envelope`.

In JDK17 it uses `com.sun.org.apache.xerces.internal.jaxp` implementation and that's work just fine and respects namespace awareness during `.reset()` call. 

Why it wasn't a problem with cling? Because this pool based mechanism was [introduced](https://github.com/jupnp/jupnp/commit/acc08768f2a18a17b1e221a7c7c12e80f7ab833c) exclusively to jupnp fork. 

I'm really not sure how this wasn't a problem for [aos-AVP](https://github.com/nova-video-player/aos-AVP/releases/tag/v6.2.63), https://github.com/nova-video-player/aos-AVP/issues/74, maybe i'm doing something wrong.

--- a/bundles/org.jupnp/src/main/java/org/jupnp/transport/impl/PooledXmlProcessor.java
+++ b/bundles/org.jupnp/src/main/java/org/jupnp/transport/impl/PooledXmlProcessor.java
@@ -32,13 +32,16 @@ import org.xml.sax.SAXException;
 
 /**
  * Abstract class responsible for creating new {@link Document}s either for writing or already parsed from a given input
- * stream. The class manages a pool ofr {@link DocumentBuilder}s internally and reuses them in different threads saving
- * cpu time from factory and builder instantiation.
- * 
+ * stream. The class manages a pool of {@link DocumentBuilder}s internally and reuses them in different threads saving
+ * CPU time from factory and builder instantiation if possible.
+ * <p>
+ * On Android this class will not reuse {@link DocumentBuilder}s because the {@link DocumentBuilder#reset()}
+ * implementation sets all internal properties to <code>false</code>
+ * <p>
  * Default pool size is 20.
- * 
- * @author Ivan Iliev - Initial contribution and API
  *
+ * @author Ivan Iliev - Initial contribution and API
+ * @author Wouter Born - Detect if pooling is possible to fix issues on Android
  */
 public abstract class PooledXmlProcessor {
 
@@ -47,6 +50,8 @@ public abstract class PooledXmlProcessor {
     private final ConcurrentLinkedQueue<DocumentBuilder> builderPool;
 
     private final transient Logger logger = LoggerFactory.getLogger(PooledXmlProcessor.class);
+
+    private boolean reuseDocumentBuilders;
 
     protected PooledXmlProcessor() {
         this(20);
@@ -57,13 +62,34 @@ public abstract class PooledXmlProcessor {
         documentBuilderFactory.setNamespaceAware(true);
         builderPool = new ConcurrentLinkedQueue<>();
 
-        for (int i = 0; i < basePoolSize; i++) {
-            try {
-                builderPool.add(documentBuilderFactory.newDocumentBuilder());
-            } catch (ParserConfigurationException e) {
-                logger.error("Error when invoking newDocumentBuilder():", e);
+        try {
+            reuseDocumentBuilders = isDocumentBuilderReusable();
+            if (reuseDocumentBuilders) {
+                logger.debug("Adding {} instances to the pool because DocumentBuilders can be reused", basePoolSize);
+                for (int i = 0; i < basePoolSize; i++) {
+                    builderPool.add(documentBuilderFactory.newDocumentBuilder());
+                }
+            } else {
+                logger.debug("Not adding instances to the pool because DocumentBuilders cannot be reused");
             }
+        } catch (ParserConfigurationException e) {
+            logger.error("Error when invoking newDocumentBuilder()", e);
         }
+    }
+
+    /**
+     * Determines if {@link DocumentBuilder} instances can be reused after resetting them.
+     * On Android it is not possible to reuse builders because {@link DocumentBuilder#reset()} sets all internal
+     * properties to <code>false</code>.
+     *
+     * @return <code>true</code> if {@link DocumentBuilder} instances can be reused after resetting them,
+     *         <code>false</code> otherwise
+     * @throws ParserConfigurationException when a {@link DocumentBuilder} cannot be created
+     */
+    private boolean isDocumentBuilderReusable() throws ParserConfigurationException {
+        DocumentBuilder builder = documentBuilderFactory.newDocumentBuilder();
+        builder.reset();
+        return builder.isNamespaceAware();
     }
 
     /**
@@ -110,13 +136,12 @@ public abstract class PooledXmlProcessor {
 
     private Document getDocument(InputSource inputSource, ErrorHandler errorHandler)
             throws FactoryConfigurationError, ParserConfigurationException, SAXException, IOException {
-        DocumentBuilder builder = builderPool.poll();
+        DocumentBuilder builder = reuseDocumentBuilders ? builderPool.poll() : null;
         if (builder == null) {
             builder = documentBuilderFactory.newDocumentBuilder();
         }
 
         try {
-
             if (errorHandler != null) {
                 builder.setErrorHandler(errorHandler);
             }
@@ -126,14 +151,15 @@ public abstract class PooledXmlProcessor {
             }
 
             return builder.newDocument();
-
         } finally {
             returnBuilder(builder);
         }
     }
 
-    private void returnBuilder(DocumentBuilder builder) throws FactoryConfigurationError, ParserConfigurationException {
-        builder.reset();
-        builderPool.add(builder);
+    private void returnBuilder(DocumentBuilder builder) throws FactoryConfigurationError {
+        if (reuseDocumentBuilders) {
+            builder.reset();
+            builderPool.add(builder);
+        }
     }
 }

